_posts/2021-05-14-Spring-Boot-TestConfiguration.md

```diff
@@ -0,0 +1,215 @@
+---
+title: Spring Boot @TestConfiguration Annotation
+categories: [spring-boot]
+date: 2021-05-15 00:00:00 +0530
+modified: 2021-05-15 06:00:00 +1000
+excerpt: "Create custom beans or override specific bean behaviors during unit testing of Spring Boot applications "
+image:
+ auto: 0035-switchboard
+---
+In this article, we will see how to create custom beans or override specific beans during unit testing by using the `@TestConfiguration` annotation.
+
+{% include github-project.html url="https://github.com/thombergs/code-examples/tree/master/spring-boot/spring-boot-testconfiguration" %}
+
+## Introducing @TestConfiguration
+
+We use the `@TestConfiguration` annotation during unit testing of Spring Boot applications for creating custom beans and/or overriding the behavior of specific beans.
+
+Let us see an example ...
+
+Let us assume that, we have a service implementation, which talks to an external REST API to get some data. The service completes its operation once the data is fetched from the API.
+
+
+The service uses [Spring WebClient](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/reactive/function/client/WebClient.html) to invoke the REST API. WebClient is a non-blocking, reactive client for making HTTP requests that works over the HTTP/1.1 protocol.
+
+During application runtime, we create the `WebClient` bean instance with appropriate configurations such as domain,port, TLS version, request/response handlers, Metrics, default headers, etc.,
+
+However, during testing, the `WebClient` is not required to be configured with such a full configuration. What the service needs is, a simple instance of `WebClient`, that is configured to talk to a mock web server.
+
+We can use `@TestConfiguration` annotation to create a `WebClient` bean with less configuration and override the original `WebClient` bean during testing.
+
+Let us create a simple version of `WebClient` bean that can be used during testing:
+
+```java
+@TestConfiguration
+public class WebClientTestConfiguration {
+ @Bean
+ public WebClient getWebClient(final WebClient.Builder builder) {
+ WebClient webClient = builder.baseUrl("http://localhost")
+ .build();
+ System.out.println("WebClient Instance Created During Testing: "
+ + webClient.toString());
+ return webClient;
+ }
+}
+```
+
+## Note on Bean Overriding
+From `Spring 5.1` onwards, the bean overriding functionality is disabled by default. A [BeanDefinitionOverrideException](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/beans/factory/support/BeanDefinitionOverrideException.html) is thrown if we attempt to override the behavior of one or more beans.
+
```

Comment:
I have rewritten this section. Have included a section to give a brief about what is bean definition overriding. 

However, I am not giving any code sample to show what is bean definition overriding. Adding a code in this section might deviate from the actual topic. In case if you think adding a code example to demonstrate the bean definition overriding, I shall add one. 

Let me know.

Reply:
You can rename this section as "Overriding Bean with TestConfiguration" to establish a connection with the main subject. Rephrase this line "When we register a bean with the spring application context, the bean will get a name". "The bean will get a name" may not be clear to the reader. A code example is not important as long as the objective of explaining the concept in simple terms is achieved. A code example is just an aid for doing that.

Reply:
I have rewritten this section. However, I am not giving any code samples, because there are different ways in which we can provide an identifier to a bean. Actually, this itself a topic to be covered in a separate blog.